### PR TITLE
Debian support

### PR DESCRIPTION
**Pull request by rodnymolina**
## Debian support effort

The following tasks have been accomplished:

1) Creation of buld-env Dockerfile for Debian-Buster.
2) Adjustment of integration-testcases to accommodate Debian's peculiarities.
3) Refactor & enhance sysbox's initialization scripts

Some more details about 3) below:

The idea here is to enhance the robustness of sysbox's initialization scripts by covering the following scenarios:

- configfs kernel module may be present as a loadable module and not loaded by default (Debian's case).
- Kernel configuration file may not be present in the expected kernel-headers location.
- unprivileged user-ns may not be enabled by default.

Signed-off-by: Rodny Molina <rmolina@nestybox.com>

### CODE REVIEW

File: scr/sysbox [hunk 1]
```diff
@@ -30,49 +30,163 @@ function retry() {
...
+
+  local distro=$1
+
+  if [[ "${distro}" == "ubuntu" ]] ||
+     [[ "${distro}" == "debian" ]]; then
+      echo "1" > /proc/sys/kernel/unprivileged_userns_clone
+
+  elif [[ "${distro}" == "centos" ]] ||
+       [[ "${distro}" == "fedora" ]] ||
+       [[ "${distro}" == "redhat" ]]; then
```
Comment (rodnymolina) [inline scr/sysbox:66]: Not quite there yet, will remove this one (redhat) in next patch.

File: scr/sysbox [hunk 2]
```diff
@@ -30,49 +30,163 @@ function retry() {
...
+
+# Identifies kernel-header's expected path based on distro.
+function kernel_header_path() {
+
+  local distro=$1
+  local path
+
+  if [[ "${distro}" == "centos" ]] ||
+     [[ "${distro}" == "fedora" ]] ||
+     [[ "${distro}" == "redhat" ]]; then
```
Comment (rodnymolina) [inline scr/sysbox:87]: Not quite there yet, will remove this one (redhat) in next patch.

File: scr/sysbox [hunk 3]
```diff
@@ -30,49 +30,163 @@ function retry() {
...
+# Ensure that kernel-modules expected by system-level apps (running within sys
+# containers) are properly loaded.
+load_required_modules() {
 
-#
-# Start Sysbox components
-#
-mkdir -p /var/lib/sysboxfs
-sysbox-mgr --log /dev/stdout > /var/log/sysbox-mgr.log 2>&1 &
+  # Sysbox requires 'configfs' module to ensure proper operation of containarized
```
Comment (rodnymolina) [inline scr/sysbox:37]: Typo, fill fix -- s/containarized/containerized

File: scr/sysbox [hunk 4]
```diff
@@ -30,49 +30,163 @@ function retry() {
...
+}
+
+# Ensures unprivileged user-ns are allowed.
+function setup_userns() {
+
+  local distro=$1
+
+  if [[ "${distro}" == "ubuntu" ]] ||
+     [[ "${distro}" == "debian" ]]; then
+      echo "1" > /proc/sys/kernel/unprivileged_userns_clone
```
Comment (ctalledo) [inline scr/sysbox:62]: In centos/fedora/redhat: is there a file that performs the same function as " /proc/sys/kernel/unprivileged_userns_clone" ?

File: scr/sysbox [hunk 5]
```diff
@@ -30,49 +30,163 @@ function retry() {
...
+
+  elif [[ "${distro}" == "centos" ]] ||
+       [[ "${distro}" == "fedora" ]] ||
+       [[ "${distro}" == "redhat" ]]; then
+
+      # Setting 10k as max user-ns value.
+      max_user_ns=$(</proc/sys/user/max_user_namespaces)
+      if [[ $max_user_ns =~ ^[0-9]+$ ]] && [[ $max_user_ns -lt 10000 ]]; then
```
Comment (ctalledo) [inline scr/sysbox:70]: Let's create a global variable for the 10K and put it at the top of this file.

File: scr/sysbox (hunk 4 above)
Comment (rodnymolina) [reply scr/sysbox:62]: Nope, there's no such a file, this is the file utilized by podman and others. I'll be updating sysbox-runc soon too.

File: scr/sysbox [hunk 6]
```diff
@@ -30,49 +30,163 @@ function retry() {
...
-  printf "\nsysbox-mgr failed to start. Here is the log file:\n"
-  cat /var/log/sysbox-mgr.log
-  exit 1
-fi
+  local kernel_path=$(kernel_header_path $distro)
 
+  if [[ ! -f "${kernel_path}"/.config ]]; then
+      # Attempt to find kernel config in /boot path.
+      if [[ -d /boot ]] && [[ -f /boot/config-$(uname -r) ]]; then
+	  cp /boot/config-$(uname -r) "${kernel_path}"/.config
```
Comment (ctalledo) [inline scr/sysbox:110]: Porque es necesario hacer esto? Que esta fallando si no lo hacemos? Yo tengo entendido que el kernel config file normalmente reside en /boot/config-*.   Me parece extrano tener que copiarlo al kernel_header_path.

File: scr/sysbox (hunk 5 above)
Comment (rodnymolina) [reply scr/sysbox:70]: Yep, like it, no literals.

File: scr/sysbox (hunk 6 above)
Comment (rodnymolina) [reply scr/sysbox:110]: Don't see any other solution for this. Without it kubeadm (or any other app looking for kernel config) will complain if this one is not found in the expected path.

File: tests/scr/sysbox [hunk 7]
```diff
@@ -28,45 +28,163 @@ function retry() {
...
-
-if [ $? -ne 0 ]; then
-  printf "\nsysbox-fs failed to start. Here is the log file:\n"
-  cat /var/log/sysbox-fs.log
-  exit 1
-fi
-
-
-exit 0
+# Ensure that kernel-modules expected by system-level apps (running within sys
```
Comment (ctalledo) [inline tests/scr/sysbox:31]: Rather than duplicating all this code, maybe we can get rid of `tests/scr/sysbox`, and instead use `scr/sysbox` but pass it a flag that tells it it's running in a test environment. If this flag is set, then we start sysbox-fs with "--ignore-handler-errors".

**Review by ctalledo — CHANGES_REQUESTED** [review]
Some comments and a change request.

File: scr/sysbox (hunk 6 above)
Comment (ctalledo) [reply scr/sysbox:110]: I am wondering if rather than making a copy at host level, we should instead have sysbox create a bind-mount from the host's `/boot/config-$(uname -r)` to the sys container's `${kernel_path}/.config`

File: scr/sysbox (hunk 5 above)
Comment (rodnymolina) [reply scr/sysbox:70]: Done.

File: scr/sysbox (hunk 6 above)
Comment (rodnymolina) [reply scr/sysbox:110]: As discussed, will push code as it is today as we're mainly dealing with a corner-case scenario (kernel config file not present in headers path after performing kernel upgrade), that appears to be affecting only Debian-Buster.

And yes, sysbox installer will be updated too.

File: tests/scr/sysbox (hunk 7 above)
Comment (rodnymolina) [reply tests/scr/sysbox:31]: Agree. Done.

File: scr/sysbox [hunk 8]
```diff
@@ -30,49 +34,175 @@ function retry() {
   false
 }
 
-# kill any running sysbox processes
-sysmgr_pids=$(pidof sysbox-mgr)
-sysfs_pids=$(pidof sysbox-fs)
+# Ensure that kernel-modules expected by system-level apps (running within sys
+# containers) are properly loaded.
+load_required_modules() {
```
Comment (ctalledo) [inline scr/sysbox:39]: nit: function load_required_modules()

Comment (rodnymolina) [reply scr/sysbox:39]: arrgh!

File: scr/sysbox [hunk 9]
```diff
@@ -6,9 +6,13 @@
 # Note: normally Sysbox is installed via a distro-specific package
 # which sets up the Sysbox systemd units. This script is meant as a
 # quick alternative to the installer to help launch Sysbox manually
-# (e.g., for testing).
+# (e.g. for development & testing).
```
Comment (ctalledo) [inline scr/sysbox:9]: Now that we've added a cmd line param for this script, let's add a comment regarding the expected invocation:

./sysbox <testing-mode>

File: scr/sysbox (hunk 8 above)
Comment (rodnymolina) [reply scr/sysbox:39]: Done.

**Review by ctalledo — APPROVED** [review]
A couple of minor comments, but looks good!

File: scr/sysbox (hunk 9 above)
Comment (rodnymolina) [reply scr/sysbox:9]: Right. Done.
